### ISSUE TITLE
Lost precision of data representation

### ISSUE THREAD

**Issue by rep-stosw** [Dobiasd/frugally-deep#434]
Hi Tobias!

I made spectral autoencoder, checked it @ Python. Then convert encoder & decoder models to JSON via frugally-deep last version.
Results are different, I was confused.

Restored spectrogram @ Python (please, click for full size) :

![Image](https://github.com/user-attachments/assets/55e9ab66-76d8-4dba-a047-bc523982da12)

Restored spectrogram @ C++ (frugally-deep).   From bottom place of spectrogram - formant lines are distorted. Also in float mode in top side some garbage was presened (please, click for full size): 

![Image](https://github.com/user-attachments/assets/da79e155-0e3b-416b-bea9-d51f73a2badc)

I tried change float/double mode and change FPU round mode - not helps.

I uploaded my Python & C++ examples/data/models/compile scripts. I hope this helps to quickly run this Issue: https://dropmefiles.com/FBGnQ

Code @ Python:

`#stub
#import os
#os.environ['TF_ENABLE_ONEDNN_OPTS'] ='0'
#os.environ['TF_CPP_MIN_LOG_LEVEL']  ='2'

from numpy import transpose 

import numpy as np
import keras

import tensorflow as tf
tf.executing_eagerly() 
tf.experimental.numpy.experimental_enable_numpy_behavior()
keras.dtype_policies.DTypePolicy(name=None)

from tensorflow.keras import layers

from keras.models import Model
from keras.layers import Input,Conv1D,MaxPooling1D,Flatten,Dense,Reshape,UpSampling1D,BatchNormalization,LeakyReLU,Conv1DTranspose,AveragePooling1D,Activation,Add,Dropout,ReLU,Subtract,LSTM,SimpleRNN,GRU
from keras.optimizers import SGD,Adam

NT=2560
Dim=64

test=np.loadtxt('spectrogram.txt',delimiter=',',usecols=range(Dim)).astype('float32')

Min=-3.7417271137237549
Max= 3.9734890460968018

Mu=Min
Sigma=Max-Min

#normalize [0,1]
test=(test-Mu)/Sigma

#limit
test=test[:NT]

test=test.reshape((len(test),Dim,1))

c_encoder=keras.saving.load_model("encoderMSE.keras",custom_objects=None,compile=False,safe_mode=True)
c_decoder=keras.saving.load_model("decoderMSE.keras",custom_objects=None,compile=False,safe_mode=True)

encoded_imgs=c_encoder.predict(test,batch_size=1)
decoded_imgs=c_decoder.predict(encoded_imgs,batch_size=1)

print(encoded_imgs.shape)
print(decoded_imgs.shape)
f=open("decoded.raw","wb")
f.write((decoded_imgs*255.0).astype('byte'))
f.close()
`

Code @ C++ :

`#define FDEEP_FLOAT_TYPE double
#include <fdeep/fdeep.hpp>

#include <stdio.h>
#include <fenv.h>

#define N 2560
#define D 64

FDEEP_FLOAT_TYPE Vectors[D*N]=   //reference spectrogram
{
 #include "spectrogram.txt"
};

FDEEP_FLOAT_TYPE VectorsP[D*N];  //predicted spectrogram

unsigned char VectorsPByte[D*N]; //grayscale [0,255]

fdeep::tensors encoded[N];

FDEEP_FLOAT_TYPE MIN=-3.7417271137237549;
FDEEP_FLOAT_TYPE MAX= 3.9734890460968018;

int main()
{
// fesetround(/*FE_DOWNWARD*/ FE_TONEAREST /*FE_TOWARDZERO*/ /*FE_UPWARD*/ );

 FDEEP_FLOAT_TYPE mu=MIN;
 FDEEP_FLOAT_TYPE sigma=MAX-MIN;

 //normalize [0,1]
 for(int i=0;i<D*N;i++)
 {
  Vectors[i]=(Vectors[i]-mu)/sigma;

  if(Vectors[i]<0.0)Vectors[i]=0.0;
  else
  if(Vectors[i]>1.0)Vectors[i]=1.0;
 }

 const auto encoder=fdeep::load_model("encoderMSE.json");
 const auto decoder=fdeep::load_model("decoderMSE.json");

 // create fdeep::tensor with its own memory
 fdeep::tensor_shape input_shape(D,1);
 fdeep::tensor input(input_shape,0.0);

 for(int n=0;n<N;n++)
 {
  for(int x=0;x<D;x++)input.set(fdeep::tensor_pos(x,1),Vectors[(D*n)+x]); //copy the values into tensor

  encoded[n]=encoder.predict({input}); //encode
 }

 for(int n=0;n<N;n++)
 {
  const auto decoded=decoder.predict({encoded[n]}); //decode

  for(int x=0;x<D;x++)VectorsP[(D*n)+x]=decoded.front().get(fdeep::tensor_pos(x,1)); //save normalized [0,1]

  //for(int x=0;x<D;x++)printf("%.3f ",VectorsP[(D*n)+x]);   //info
 }

 //convert to byte
 for(int i=0;i<D*N;i++)
 {
  FDEEP_FLOAT_TYPE data=VectorsP[i];

  if(data<0.0)data=0.0;
  else
  if(data>1.0)data=1.0;

  VectorsPByte[i]=(unsigned char)(data*255.0);
 }

 FILE *f=fopen("decoded.raw","wb");
 fwrite(VectorsPByte,1,D*N,f);
 fclose(f);

//------------

 //convert to byte
 for(int i=0;i<D*N;i++)VectorsPByte[i]=(unsigned char)(Vectors[i]*255.0);

 f=fopen("input.raw","wb");
 fwrite(VectorsPByte,1,D*N,f);
 fclose(f);

//------------

 return 0;
}
`

Please, help me.  How to make precision/bitexact (remove spectral distortions) ?

**Comment by Dobiasd** [comment]
Hi, and thanks for the interesting way to compare the results between Keras and frugally-deep.

I downloaded the `_for_test_.zip_pass_123.zip` file, but it requires a password to extract. What is it?

**Comment by Dobiasd** [comment]
Oh, nevermind. The password is `123`, as the filename says. 😁

I'll look into the problem and get back to you here.

**Comment by Dobiasd** [comment]
Without deeper digging, it's hard for me to see the intended equivalence between the Python code and the C++ code.

There are a lot of things going on in addition to the actual forward pass of the models.
In the C++ code, for example, values are clamped into the interval `[0.0, 1.0]`, but I don't see this happening in the Python code.

Could you please reduce things to a more minimal example, one that shows that for the same input values, a prediction in frugally-deep yields a result different from one in Keras with the same inputs?

Something like the following would be great:

```python3
import keras
import numpy as np

model = keras.saving.load_model("decoderMSE.keras", compile=False)
input = np.array([[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]])
print(input)
print(input.shape)
result = model.predict([input])
print(result)
print(result.shape)
```

```cpp
#include <fdeep/fdeep.hpp>

int main() {
    const auto model = fdeep::load_model("decoderMSE.json");
    const fdeep::tensor input = {
        fdeep::tensor(fdeep::tensor_shape(static_cast<std::size_t>(8)),
        std::vector<float>{
            0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8
        }),
    };
    std::cout << fdeep::show_tensor(input) << std::endl;
    std::cout << fdeep::show_tensor_shape(input.shape()) << std::endl;
    const auto result = model.predict({input});
    std::cout << fdeep::show_tensor_shape(result.front().shape()) << std::endl;
    std::cout << fdeep::show_tensors(result) << std::endl;
}
```

Of course replace `0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8` with the values you get a difference with. And in case it's not with `decoderMSE`, but with `encoderMSE`, of course, use this.

But having it narrowed down like this would be extremely helpful.

**Comment by rep-stosw** [comment]
> Oh, nevermind. The password is `123`, as the filename says. 😁

I can't do anything about it. It puts a password on the archive itself when I uploaded it.


> Could you please reduce things to a more minimal example, one that shows that for the same input values, a prediction in frugally-deep yields a result different from one in Keras with the same inputs?

Problem solved.  This was my mistake. I put index 1 instead 0 in tensor_pos(), so data was shifted. Corrected code fragment in C++ part:

`  for(int x=0;x<D;x++)input.set(fdeep::tensor_pos(x,0),Vectors[(D*n)+x]); //copy the values into tensor`

`  for(int x=0;x<D;x++)VectorsP[(D*n)+x]=decoded.front().get(fdeep::tensor_pos(x,0)); //save normalized [0,1]`

Now output result is correct.

Thanks for help.  This issue can be closed already.

**Comment by Dobiasd** [comment]
Great! 🎉
Glad you figured it out. 👍
